### ISSUE TITLE
Is there an equivalent to shouldRasterizeIOS?

### ISSUE THREAD

**Issue by msageryd** [react-native-community/react-native-svg#573]
Svg.Text performs quite bad when transforming the `<Svg>`. Is there any way to rasterize the text as with the `shouldRasterizeIOS` property on a regular `<View>`?

**Comment by msand** [comment]
Have you tried doing the transformation on a containing react-native view instead? Keep the svg unchanged, and, ensure it either doesn't re-render or at least uses the cached output of the text rendering.

**Comment by msageryd** [comment]
I'm can't make that work in my setup. If I knew it would give much better performance I'd put some more work into it.

This is what I do (simplified):

    <View  (screen sized)>
      <Svg (screen sized)>
        <G ref={ref => (this.svgGroupRef = ref)}>
          {this.shapes.map(shape => {
            if (!shape.isVisible) return null;
            return <MyShape />;
          })}
        </G>
      </Svg>
    </View>;

- MyShape does never re-render (shouldComponentUpdate => false)
- Since <G> does not have a dimension per se, it can live inside of a screen sized Svg and still scale infinitely with vector resolution.
- All transforms are done with setNativeProps

The problem with transforming the outer view is that I'd transform the whole Svg as well. The Svg component seems to allocate memory in accordance to it's size. <G> on the other hand does not (G seems to behave about the same as ART.Surface in that regard)

In my example I have about 1000 shapes. They are quite simple, a rect with round corners a line from the rect and a number inside of the rect. Before rendering everything I'm calculating the boundingRect for all shapes so I can quickly decide if a shape is visible or not.

If I skip the text rendering I get good performance. But as soon as I render the text I get very poor  performance.

What do you mean by "cached output of text rendering"? This seems to be exactly what I have looked for. How do I accomplish this?

**Comment by msand** [comment]
@msageryd Put a breakpoint here:
https://github.com/react-native-community/react-native-svg/blob/master/ios/Text/RNSVGTSpan.m#L71-L73
and here: https://github.com/react-native-community/react-native-svg/blob/master/android/src/main/java/com/horcrux/svg/TSpanShadowNode.java#L74-L76
and make sure it uses the cached path, or that the sub-tree containing it isn't invalidated.

**Comment by msageryd** [comment]
Thanks.
I could debugged this in XCode (iOS simulator).

To ensure that I hadn't goofed I put a render-counter in the components render function. I can confirm that the component is only rendered one time.

I've simplified the component so the only thing inside of my `<G>` is a `<Text>` with the string "1".

I'm passing the `if (_cache) {..` four times as per the below:
1. _cache === false
2. _cache === true
3. _cache === false
4. _cache === true

These four passes happens when I first render the view. On each following setNativeProps (only X- and Y translation) I pass this section twice. First with no cache and then with _cache === true.

**Comment by msand** [comment]
Are you setting native props on the text element? Have you tried only setting it on the G element? Does it still invalidate the text cache?

**Comment by msageryd** [comment]
No, I only set it on the G component. The whole purpose is to proportionally transform all components within the G.

**Comment by msand** [comment]
I haven't actually looked at the invalidation nor verified that it is only done when needed. You could set breakpoints in all the places where invalidate is called, or the cached path release, and try to find out why it keeps re-rendering.

**Comment by msageryd** [comment]
Thanks.
I'm trying hard to follow the code (I've never looked at objC before).

Tried to compare TSpan with Rect, but now I understand even less. 

- Rect gets invalidated on every change in x, y, width, etc and has no _cache.
- TSpan gets invalidated on setContent() and has some magic autorelease for _cache.

I assume that setNativeProps will bypass setX, setWidth etc on Rect? Where should I look in the code to try to understand where setNativeProps does it's magic? One idea on the bad performance could be that Text doesn't share the same behaviour regarding setNativeProps, hence re-renders every time.

**Comment by msand** [comment]
There are probably some optimization opportunities here. Could split the calculation of the path for the elements, and the composition of the matrix with the path, to render it to the canvas. Such that if only the matrix transformation changes, it doesn't need to recalculate what to render, only the re-render the same content with a possibly different transform. This would also complicate the logic a bit, but essentially, having two different invalidation handlers, one for transforms (doesn't clear the cached path), and one for changes which can affect the content (clears the cached path), should speed up simple transform animations significantly.
And, looking at the code now, I'm starting to think the cached text paths are always released no matter what right now. I would have to step through the relevant logic to verify. I'm sure there's lots of low hanging fruit here.

**Comment by msageryd** [comment]
Wow! That seems really great. Let me know if I can help out with something.

**Comment by msageryd** [comment]
Would your proposed solution make my use of setNativeProps obsolete? If that would be the case I could switch to PureComponent instead of Component. As of now I'm prohibiting component updates via shouldComponentUpdate in the case when I only want to change the matrix (i.e via setNativeProps).

**Comment by msand** [comment]
Well, using PureComponent would probably simplify reasoning about your code at least, but limiting updates and diffing, e.g. if you know that only exactly one property has changed on a single node, such as the matrix in this case, it is probably still faster to make the bridge set that single property directly on the node using the setNativeProps, and skip the lifecycle with all its property extraction, transformation, diffing, etc. 

But, making sure that all nodes cache their paths, and reuse that if only the matrix changes, should certainly make rendering more fluid in both cases. And, correctly used data structures and property passing, such that all the PureComponents only ever need to do a single reference equality check each, for a use case dependent splitting of responsibilities into composed components and according to their relative update frequencies, should make the ux good enough for some unknown larger amount of rendering work.

Anyway, intelligently applied mutability can certainly speed up the hottest parts of your render loop, usually with a trade-off in reasoning complexity, in any single threaded von neumann architecture at least ;) In a more parallel (eg. opengl) or distributed processing unit/environment you probably want to avoid mutating shared state and synchronization as much as possible instead. And, if you need to render large amounts of work, you should probably look at moving over to the accelerated gl hardware asap.

**Comment by msand** [comment]
I thing which would be very helpful is if you could make some representative performance tests that we can use to compare the effects of various changes. Alternatively, ensure the all nodes cache the path, invalidate if any content affecting property changes, cleanup correctly when detached, and make sure that changing the matrix property only causes invalidation of the hitareas and a rerender, no invalidation of the cached paths.

**Comment by msageryd** [comment]
Ok, I'm with you regarding the performance tests. I could try to whip up something. Do you know of if there are any APIs in RN for measuring actual fps? I'm thinking of making a simple app with a bunch of Rect and Text elements which are animated over the screen, but I don't know how to measure performance objectively.

The other stuff (ensure paths are cached etc) is out of my league. I've never worked with objectiveC.

**Comment by msand** [comment]
Ok, sounds good. I haven't actually spent almost any time on optimizing nor profiling React-Native yet, I would have to google around a bit first for the state-of-the-art. The next sprint at work will finally have some time allocated for this kind of work, so at that point at the latest I'll check the available tools. If rnsvg shows up as the top priority bottle-neck, or if we manage to fix any other potentially more fruitful optimizations first, then I'll be able to put something else than my free-time on this.

**Comment by msand** [comment]
@msageryd I've now made some changes to the caching of paths, but I don't have any good performance tests yet so it hasn't been tested much at all. You can check it out here https://github.com/react-native-community/react-native-svg/pull/596
The changes were relatively minor in the end:
https://github.com/react-native-community/react-native-svg/pull/596/commits/149f4601084a4ba3cec56844e4c91ff4f2c4ef62
https://github.com/react-native-community/react-native-svg/pull/596/commits/db8f1da46b3439435f6b057897936955d229f4bb
The branch also supports nesting View inside Svg, and Svg inside Svg, the changes from this might have some performance penalty as well (at least on android).

**Comment by msageryd** [comment]
@msand Sorry for taking long time to answer. I've been away skiing. 

This is awesome news. I will try it out first thing next week. I will also try to make time for some kind of objective performance testing.

**Comment by msand** [comment]
While testing it in a project, I noticed some regressions with the transforms, something stays stale in the cache, so the current work in progress isn't ready for production at least. Any help with testing much appreciated.

**Comment by msageryd** [comment]
@msand Just an update.. I've been overloaded with work and I haven't had time to do any tests. This topic is very important to me, so I'll give it a good test for sure. I'm hoping to be out of this work load in a week or two.

**Comment by msand** [comment]
Alright, looking forward to it 😃 The issues were caused by using texture view on android, so i reset the branch to before that and merged it in. The optimizations are available in the latest version.

**Comment by msageryd** [comment]
@msand I upgraded from 6.2.2. to 6.3.1. There is something really fishy with text rendering. When I move around my shapes (dragging with fingers) the text is rendered at wrong positions (i.e. out of whack) on iOS, but it works as it should in the simulator. I don't know where to start looking for the problem. 

Would you like me to send you a screen recording from the app before /after upgrade?

**Comment by msageryd** [comment]
I recorded a screen capture. When I did this I noticed a pattern. It seems like some text is transformed along the wrong axis. Example: I have a rect with text inside. When I move the rect upwards (negative y), the text moves to the left (negative x) and vice versa. This occurs with text that is transformed (rotated). Straight (non rotated) text stays in one spot, i.e. it does not move with the rect at all.

Please tell me if you want to take a look at the recordings.

**Comment by msageryd** [comment]
I managed to make some gifs. The first one is with RNSVG 6.2.2. Notice  the number in the rectangle as well as the measurement along the grey line.

![6 2 2](https://user-images.githubusercontent.com/4483085/38132024-a0bdf49e-3409-11e8-8d42-2be8a89bce65.gif)

The exact same codebase but with RNSVG 6.3.1. Notice that the number in tre rectangle is stays in it's original position, i.e. it's not transformed at all. The measurement is transformed along the wrong axis.

![6 3 1](https://user-images.githubusercontent.com/4483085/38132027-a4a267d4-3409-11e8-98d4-72c79f71b0d2.gif)

**Comment by msageryd** [comment]
I'm back to this issue after working on some other stuff. There are two problems. I think that #2 is kind of the same issue as #709. 

1. I'll go on and try to bisect the problem with the seemingly exchanged axis on the rotated text element.

2. I have zoomed in on the problem with the missing re-rendering of the text in the rect. It turns out that the `<Text>` element does not trigger a re-render when the x- or y-props are changed. This worked better in older versions. An ugly temporary solution is to assign a unique `key` property to the text.

**Comment by msageryd** [comment]
I have now bisected the commits. This took some work, because not every committed version was runnable. I'm down to two suspects, which is as far as I can get.

**Working as expected:**
The last working commit was **149f460 - Improve path cashing on android**. 

**Crashing:**
The next commit is not runnable, **db8f1da - Improve path caching ios**. As soon as an SVG element should render the app freezes, sometimes with a black screen.

**Not working as expected:**
The commit after this is **cd43fc1 - Optimize hit testing**. This commit has the two bugs shown in my gifs above.

**Conclusion:**
The bug was introduced in either db8f1da or cd43fc1.

@msand do you get any clues from this?
Should I post this as a separate issue, or keep it in this thread?

**Comment by msand** [comment]
At least one of the issue lies with the path caching. Would need to add a recursive releaseCachedChildPaths method to RNSVGContainer, traversing the tree while there's any cached paths, such that any nodes which depend on updated parent properties can rerender, or revert the caching logic altogether.

**Comment by msageryd** [comment]
Regarding reverting the caching logic: 

If this is avoidable it would be good. Text rendering was suffering hard from excessive redrawing before your caching fix. I haven't had time for any performance tests yet, though. So I actually don't know how much the caching brings in terms of FPS.

**Comment by msand** [comment]
Btw, thanks for bisecting and narrowing down the issue. We can probably keep the discussion here for now, as the earlier messages are somewhat related to it anyway. But feel free to open separate issues if you feel like it, or come up with even better ways to describe them, it might help others investigate as well. Especially if you can make minimal reproductions, that would be great 👍
Regarding the text elements and x/y props, are you changing them using setNativeProps or using normal prop diffing?
Can you try reverting the change made to RNSVGNode.m in https://github.com/react-native-community/react-native-svg/commit/db8f1da46b3439435f6b057897936955d229f4bb#diff-fd13a4ce6718d799f74859c5145670a3L133, specifically the method: 
```obj-c
 - (void)setMatrix:(CGAffineTransform)matrix
```
such that you call 
```obj-c
    [self invalidate];
```
instead of 
```obj-c
    id<RNSVGContainer> container = (id<RNSVGContainer>)self.superview;
    [container invalidate];
```
It's a bit hard to figure out your current setup without access to the code, and as I don't have any way to replicate the issue at the moment I'm not sure how to proceed either. Have you tried with the latest version btw?

**Comment by msand** [comment]
@msageryd Published v7.0.0 now 😄

**Comment by msand** [comment]
@msageryd Thought you might be interested in a very successful optimisation strategy I applied the other day. It works great to speed up animations/gestures where you're only transforming some parts of the content, or where only some parts of the svg content rarely but various layers need transforms and clipping. 

It roughly goes like this: Cover the area you want to respond to gestures with a Animated.View/ScrollView, split the svg content you want to animate independently into separate svg roots and wrap them in a Animated.View, then connect the gestures to transform the entire results of rendering the svg content into bitmaps using RN transform arrays and useNativeDriver: true. 

This way the svg doesn't need to rerender the bitmap at all, and it uses native Views to clip and transform the content instead.
https://snack.expo.io/@msand/soundcloud-player-fast-animation-cross-platform

There's a bit more info on medium: https://medium.com/@dnasleakim/the-fast-cross-platform-end-result-of-my-investigation-can-be-found-here-ed2f0ac79ac7

Expo required a bit of extra logic for Android. But the iOS version seems to work fine in the latest rn and rnsvg.
https://medium.com/@dnasleakim/and-for-the-grand-finale-the-fastest-possible-i-think-8eccf71f6e2c

Might be easier to follow the chain of optimisations in chronological order on my profile, as the responses became a bit of a tree over time.
https://medium.com/@dnasleakim/responses

E.g. the first change I made was based on looking at the time profile and noticing the endTransparencyLayer was the main time consumer, so I moved the opacity from being on thousands of rects onto a single G element instead, making it already quite usable. 

Then, trying to make the bitmap smaller seemed to improve a bit, but the best was just not to rerender any bitmaps at all (especially screen sized), merely clipping and transforming the view of the bitmaps, invalidating the dirty areas and as little extra as possible. 

In practice, this runs the whole interaction of the app using only native code, the js is only used to set up the native view hierarchy which then does the initial single render of the svgs, sets up the native gesture responders and animations, and gives full 60fps interaction.

**Comment by msageryd** [comment]
@msand, wow thanks!
This is very interesting. For now I have *ok* performance, but I will need to increase the speed when we take on larger projects (i.e more svg shapes). It would also be good with a bit more speed for users with older devices.

I'll dig into this, for sure.
